File: src/core/project.cpp
#include "project.h"
#include <set>

namespace sugar {

Project::Project(const Config& config, const std::filesystem::path& root_dir)
    : config(config), root_dir_(root_dir) {
}

std::vector<std::filesystem::path> Project::getSourceFiles() const {
    std::vector<std::filesystem::path> source_files;
    std::set<std::string> source_extensions = {".cpp", ".cc", ".cxx", ".c", ".h", ".hpp"};

    for (const auto& src_path : config.source_paths) {
        std::filesystem::path src_dir = root_dir_ / src_path;
        if (!std::filesystem::exists(src_dir)) {
            continue;
        }

        // Search only in the specified directory (not subdirectories)
        for (const auto& ext : source_extensions) {
            for (const auto& entry : std::filesystem::directory_iterator(src_dir)) {
                if (entry.is_regular_file() && entry.path().extension() == ext) {
                    source_files.push_back(entry.path());
                }
            }
        }
    }

    return source_files;
}

std::filesystem::path Project::getBuildDirectory() const {
    return root_dir_ / config.build_path;
}

std::filesystem::path Project::getOutputDirectory() const {
    return root_dir_ / config.output_path;
}

std::string Project::getTargetFilename() const {
    std::string target_name = config.project_name;

    if (config.platform == "Windows") {
        if (config.project_type == "exe") {
            return target_name + ".exe";
        } else if (config.project_type == "static") {
            return target_name + ".lib";
        } else if (config.project_type == "shared") {
            return target_name + ".dll";
        }
    } else if (config.platform == "Linux") {
        if (config.project_type == "exe") {
            return target_name;
        } else if (config.project_type == "static") {
            return "lib" + target_name + ".a";
        } else if (config.project_type == "shared") {
            return "lib" + target_name + ".so";
        }
    } else if (config.platform == "macOS") {
        if (config.project_type == "exe") {
            return target_name;
        } else if (config.project_type == "static") {
            return "lib" + target_name + ".a";
        } else if (config.project_type == "shared") {
            return "lib" + target_name + ".dylib";
        }
    }

    return target_name;
}

} // namespace sugar

File: python version/core/project.py
"""Project representation and management."""

from pathlib import Path
from typing import List
from .config import Config


class Project:
    """
    Represents a SugarBuilder C++ project.
    
    Manages project configuration, source files, and build artifacts.
    """
    
    def __init__(self, config: Config, root_dir: str | Path = "."):
        """
        Initialize project with configuration.
        
        Args:
            config: Project configuration from sugar.toml.
            root_dir: Root directory of the project.
        """
        self.config = config
        self.root_dir = Path(root_dir)
    
    def get_source_files(self) -> List[Path]:
        """
        Collect all C++ source files from configured source paths.
        
        Returns:
            List of Path objects pointing to C++ source files (.cpp, .cc, .cxx, .c, .h, .hpp).
        """
        source_files = []
        # Include both source and header files from source_paths
        # Headers in source_paths are included during compilation
        source_extensions = {".cpp", ".cc", ".cxx", ".c", ".h", ".hpp"}
        
        for src_path in self.config.source_paths:
            src_dir = self.root_dir / src_path
            if not src_dir.exists():
                continue
            
            # Search only in the specified directory (not subdirectories)
            for ext in source_extensions:
                source_files.extend(src_dir.glob(f"*{ext}"))
        
        return source_files
    
    def get_build_directory(self) -> Path:
        """
        Get the build directory path.
        
        Returns:
            Path to the build directory.
        """
        return self.root_dir / self.config.build_path
    
    def get_output_directory(self) -> Path:
        """
        Get the output directory path.
        
        Returns:
            Path to the output directory.
        """
        return self.root_dir / self.config.output_path
    
    def get_target_filename(self) -> str:
        """
        Get the target executable/library filename based on project type and platform.
        
        Returns:
            Filename of the build target.
        """
        target_name = self.config.project_name
        
        if self.config.platform == "Windows":
            if self.config.project_type == "exe":
                return f"{target_name}.exe"
            elif self.config.project_type == "static":
                return f"{target_name}.lib"
            elif self.config.project_type == "shared":
                return f"{target_name}.dll"
        
        elif self.config.platform == "Linux":
            if self.config.project_type == "exe":
                return target_name
            elif self.config.project_type == "static":
                return f"lib{target_name}.a"
            elif self.config.project_type == "shared":
                return f"lib{target_name}.so"
        
        elif self.config.platform == "macOS":
            if self.config.project_type == "exe":
                return target_name
            elif self.config.project_type == "static":
                return f"lib{target_name}.a"
            elif self.config.project_type == "shared":
                return f"lib{target_name}.dylib"
        
        return target_name
